### PR TITLE
feat: lora fine-tuning in FHE + gpt2 use case example

### PR DESCRIPTION
**Pull request by RomanBredehoft**
currently running the notebook with 100 epochs

refs https://github.com/zama-ai/concrete-ml-internal/issues/4522

### CODE REVIEW

File: src/concrete/ml/torch/lora.py [hunk 1]
```diff
@@ -0,0 +1,261 @@
+"""This module contains classes for LoRA (Low-Rank Adaptation) training and custom layers."""
+
+import torch
+
+# pylint: disable=abstract-method
+# pylint: disable=arguments-differ
+
+
+class LoraTraining(torch.nn.Module):
```
[the remark this answers is not in the document]
Comment (jfrery) [reply src/concrete/ml/torch/lora.py:9]: that's a bit redundant. lora stands for low rank adaptation.

File: use_case_examples/lora_finetuning/README.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+This project demonstrates how to fine-tune GPT-2 using Low-Rank Adaptation (LoRA) weights with Fully Homomorphic Encryption (FHE). The goal is to train a specialized model in a privacy-preserving manner, with minimal memory requirements.
+
+## Overview
+
+Fine-tuning large language models typically requires access to sensitive data, which can raise privacy concerns. By leveraging FHE, we can perform computations on encrypted data, ensuring that the data remains private throughout the training process. In this approach, the LoRA weights are only known to the user who owns the data and the memory hungry foundation model remains on the server.
+
+## Key Features
+
+- **LoRA Fine-Tuning**: Fine-tune GPT-2 by adapting low-rank weights.
```
[the remark this answers is not in the document]
Comment (jfrery) [reply use_case_examples/lora_finetuning/README.md:11]: Well that's the gpt2 use case.

File: src/concrete/ml/torch/lora.py [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+        self.optimizer = None
+        self.lr_scheduler = None
+
+        self.gradient_accumulation_steps = gradient_accumulation_steps
+        self.max_grad_norm = None
+
+        self.calibrate = False
+        self.run_optimizer = False
+
+    def _validate_model_type(self):
```
[the remark this answers is not in the document]
Comment (jfrery) [reply src/concrete/ml/torch/lora.py:31]: Haven't tested with other models yet.

File: src/concrete/ml/torch/lora.py [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+
+    SUPPORTED_MODELS = ["gpt2"]
+
+    def __init__(self, inference_model, gradient_accumulation_steps) -> None:
+        super().__init__()
+
+        self.inference_model = inference_model
+
+        # Validate the base model type
+        self._validate_model_type()
```
[the remark this answers is not in the document]
Comment (jfrery) [reply src/concrete/ml/torch/lora.py:20]: I didn't test it yet but we will adapt it to MLP and llama yes

File: use_case_examples/lora_finetuning/README.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+- **Fully Homomorphic Encryption**: Perform training and inference on encrypted data.
```
[the remark this answers is not in the document]
Comment (jfrery) [reply use_case_examples/lora_finetuning/README.md:12]: Alright

File: src/concrete/ml/torch/lora.py [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+                )
+
+        except AttributeError as e:
+            raise ValueError(
+                "Unable to determine the base model type. "
+                "Ensure that the inference_model has a "
+                "'base_model.model.config.model_type' attribute."
+            ) from e
+
+    def update_training_parameters(self, optimizer, lr_scheduler, training_args):
```
[the remark this answers is not in the document]
Comment (jfrery) [reply src/concrete/ml/torch/lora.py:57]: Not sure. that's how it's been done so far.

File: src/concrete/ml/torch/lora.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jfrery) [reply src/concrete/ml/torch/lora.py:9]: LoraFHEmodel?

File: src/concrete/ml/torch/lora.py [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+            The gradient input tensor after applying the backward pass.
+        """
+        backward_module = ctx.backward_module
+        grad_input = backward_module.forward(grad_output)
+
+        # grad_weight and grad_bias are not needed when computing the backward for LoRA
+        return grad_input, None, None
+
+
+def get_remote_names(model: torch.nn.Module, include_embedding_layers: bool = False) -> List[str]:
```
Comment (andrei-stoian-zama) [inline src/concrete/ml/torch/lora.py:296]: this is too specific for a library function. e.g. for the `lm_head` moniker. 

I suggest you add two arguments `remote_layer_types` and `layer_reject_filter`. 

You can then call with `remote_layer_types = [nn.Linear, nn.Embedding, transformer.Conv1D]`. Thus CML does not need to know `TransformerConv1D`. and `layer_reject_filter = ['lm_head`]` 

You can thus remove the `include_embedding_layers` flag.

Comment (andrei-stoian-zama) [reply src/concrete/ml/torch/lora.py:296]: but in the function we manually add `CustomLinear` if the user asks for `nn.Linear`.

**Review by andrei-stoian-zama — CHANGES_REQUESTED** [review]
I would like to remove the transformer dependency:

- move the dep to the use case example, replace conv1d -> nn.Linear in the usecase 
- parametrize the remote module finding function in way that avoids the dep

File: src/concrete/ml/torch/lora.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrei-stoian-zama) [reply src/concrete/ml/torch/lora.py:12]: we'll leave this for the later refactoring

File: src/concrete/ml/torch/lora.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (andrei-stoian-zama) [reply src/concrete/ml/torch/lora.py]: we'll move these in the refactoring

File: src/concrete/ml/torch/lora.py [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+            training_args: The training arguments containing gradient
+                accumulation steps and max grad norm.
+        """
+        assert self.gradient_accumulation_steps == training_args.gradient_accumulation_steps
+
+        self.optimizer = optimizer
+        self.lr_scheduler = lr_scheduler
+        self.max_grad_norm = training_args.max_grad_norm
+
+    def forward(self, inputs):
```
[the remark this answers is not in the document]
Comment (andrei-stoian-zama) [reply src/concrete/ml/torch/lora.py:107]: we'll discuss this again for the refactoring

File: tests/torch/test_lora.py [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+
+    expected_grad_input = grad_output @ weight.t()
+    assert grad_input is not None and torch.allclose(grad_input, expected_grad_input)
+
+
+def test_custom_conv1d():
+    """Test the CustomConv1D module."""
+    weight = torch.tensor([[1.0, 2.0], [3.0, 4.0]], requires_grad=True)
+    bias = torch.tensor([0.5, -0.5], requires_grad=True)
+    module = CustomConv1D(weight, bias)
```
[the remark this answers is not in the document]
Comment (andrei-stoian-zama) [reply tests/torch/test_lora.py]: ok, maybe not a good idea to move to the lib to avoir `transformers` dependency.

File: tests/torch/test_lora.py [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+
+from collections import namedtuple
+from types import SimpleNamespace
+
+import pytest
+import torch
+from torch import nn
+from torch.optim import SGD
+from torch.optim.lr_scheduler import StepLR
+from transformers import Conv1D as TransformerConv1D
```
Comment (andrei-stoian-zama) [inline tests/torch/test_lora.py:13]: this is ok to have here, transformers can be a dev dependency

File: src/concrete/ml/torch/lora.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jfrery) [reply src/concrete/ml/torch/lora.py]: removed

File: src/concrete/ml/torch/lora.py [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+
+import torch
+
+# pylint: disable=abstract-method
+# pylint: disable=arguments-differ
+
+
+class LoraTraining(torch.nn.Module):
+    """LoraTraining module for fine-tuning with LoRA."""
+
```
[the remark this answers is not in the document]
Comment (jfrery) [reply src/concrete/ml/torch/lora.py:14]: Added

File: src/concrete/ml/torch/lora.py (hunk 7 above)
Comment (jfrery) [reply src/concrete/ml/torch/lora.py:309]: I am not sure here. This is very specific with the `replace_layers_with_custom`. I would not allow the user to select layers himself. For now we do linear layers in FHE and we have a workaround for lm_head and embedding as long as there are not fixed.

File: src/concrete/ml/torch/lora.py [hunk 12]
```diff
@@ -0,0 +1,326 @@
+"""This module contains classes for LoRA (Low-Rank Adaptation) training and custom layers."""
+
+from typing import List
+
+import torch
+from transformers import Conv1D as TransformerConv1D
```
[the remark this answers is not in the document]
Comment (jfrery) [reply src/concrete/ml/torch/lora.py]: alright done

**Review by andrei-stoian-zama — APPROVED** [review]
Excellent work here @RomanBredehoft and @jfrery !

**Pull request merged**